### PR TITLE
Fix automatic optimizer LR with DDP training

### PR DESCRIPTION
**Pull request by Y-T-G**
`set_model_attributes()` adds some attributes to the model before training starts including `nc`:
https://github.com/ultralytics/ultralytics/blob/34b339d0331314808d6564a0213e6b3bc39ff613/ultralytics/engine/trainer.py#L236
https://github.com/ultralytics/ultralytics/blob/34b339d0331314808d6564a0213e6b3bc39ff613/ultralytics/models/yolo/detect/train.py#L81

However, in DDP training, the model is wrapped again with DDP which removes these attributes:
https://github.com/ultralytics/ultralytics/blob/34b339d0331314808d6564a0213e6b3bc39ff613/ultralytics/engine/trainer.py#L272-L273

So when the optimizer is build, it can't find `nc` and uses `nc=10` as fallback:
https://github.com/ultralytics/ultralytics/blob/34b339d0331314808d6564a0213e6b3bc39ff613/ultralytics/engine/trainer.py#L784

The result is a different LR when launching single-GPU training vs. DDP:

**Single-GPU**
```bash
yolo train model=yolo11n.pt data=coco128.yaml epochs=30 device=0

optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000119, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
```

**Multi-GPU**
```bash
yolo train model=yolo11n.pt data=coco128.yaml epochs=30 device=0,1

optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.000714, momentum=0.9) with parameter groups 81 weight(decay=0.0), 88 weight(decay=0.0005), 87 bias(decay=0.0)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary  
Improves Distributed Training setup by ensuring model attributes are correctly configured after applying DDP (DistributedDataParallel). 🚀  

### 📊 Key Changes  
- Added `self.set_model_attributes()` call after wrapping the model in `DistributedDataParallel` (DDP).  

### 🎯 Purpose & Impact  
- 🛠 **Purpose**: Ensures model attributes are properly configured after DDP wrapping, avoiding potential issues with model behavior during distributed training.  
- 🌐 **Impact**: More stable and reliable behavior for users running distributed training setups, especially in complex scenarios.  
- 💡 **Benefit**: Reduces buggy behavior and potential misconfigurations, leading to smoother multi-GPU training experiences.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

### 🐞 Bug Context
It seems that your PR addresses an issue with learning rate discrepancies during single-GPU vs. DDP training due to the removal of attributes in the DDP wrapping process. If this resolves a discovered bug, great work! 🎉

#### Request for Additional Context:
- If you haven't already, could you please provide a **minimum reproducible example (MRE)** to ensure we can efficiently test and validate the issue? This will help streamline the debugging and testing process.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don’t hesitate to leave a comment if you have any questions. This is an automated response to help improve your contribution workflow 🚀—an Ultralytics engineer will review your PR shortly. Thank you for contributing to Ultralytics! 🙌

**Comment by Y-T-G** [comment]
@Laughing-q This PR affects DDP training. So needs to be tested.

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Y-T-G, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! Your fix for maintaining model attributes after DDP wrapping enhances distributed training reliability, which is an important improvement. To ensure a seamless integration, please review the following checklist:

- ✅ **Define a Purpose**: Your PR description is detailed and clearly outlines the purpose and impact of your changes. If there are any associated [issues](https://github.com/ultralytics/ultralytics/issues), linking them would be helpful for context.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind the latest updates, please synchronize by clicking 'Update branch' or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks pass. If any checks fail, kindly address them to ensure alignment with our quality standards.
- ✅ **Update Documentation**: If your changes affect behavior or functionality, please update the corresponding [documentation](https://docs.ultralytics.com) so that users benefit from your enhancement.
- ✅ **Add Tests**: If your changes can be tested, ensure that appropriate tests are included or updated to reflect your modifications, and verify that all tests pass.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first contribution. You can confirm this by writing "I have read the CLA Document and I sign the CLA" as a new comment.
- ✅ **Minimize Changes**: Confirm that your changes are limited to what's necessary for your fix. Simplifying contributions helps maintain focus and ensures seamless reviews.

For more information, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). 🤝

This is an automated response to guide you with our PR process. An Ultralytics engineer will review your contribution shortly and might provide further feedback. Thank you for helping improve Ultralytics! 🚀

**Comment by glenn-jocher** [comment]
@Y-T-G ah great catch! @Laughing-q please test.

**Comment by glenn-jocher** [comment]
Merging now, we can rework later if issues arise.

**Comment by UltralyticsAssistant** [comment]
🎉 Your PR has been merged! Huge thanks to @Y-T-G for leading this improvement and to @glenn-jocher for contributing your expertise. This enhancement ensures smoother, more reliable distributed training—an awesome step forward for our users! 🚀  

As Marcus Aurelius once said, *"What stands in the way becomes the way."* By addressing potential challenges in DDP, you’ve paved the path for more seamless multi-GPU training. Your work truly makes a difference. Thank you for your dedication and thoughtful contributions! 💡✨

**Comment by Laughing-q** [comment]
@Y-T-G Can we directly use the `self.data["nc"]` here for this?

**Comment by Y-T-G** [comment]
@Laughing-q Yes

**Comment by Laughing-q** [comment]
@Y-T-G ok got it! submitted a PR here: https://github.com/ultralytics/ultralytics/pull/18842
